### ISSUE TITLE
github-merge: context not isolated

### ISSUE THREAD

**Issue by real-or-random** [bitcoin-core/bitcoin-maintainer-tools#49]
One potential pitfall of github-merge is that works in your normal workspace. That is, you can switch branches, change things, etc, between two git commands issued by the script, in particular when it waits for user input. For example, if you switch to a different branch and then type `s`, you sign the wrong commit. (I haven't tested it but this is what I assume from reading the code.)

Could git worktrees help here?
https://git-scm.com/docs/git-worktree

**Comment by laanwj** [comment]
When it throws you into a shell there is code to guard against this ("ERROR: Tree hash changed unexpectedly"), but right, not when it's waiting for `s`. I think you could force it into making a mistake if you switch branches in a different terminal window.

Could add a further check that the `git commit --gpg-sign` only signs the previously determined top commit. I wouldn't mind some more belts and suspenders here, though I don't think it's possible to protect against a malicious parallel usage of git in general (there's always some window for a race condition).

A worktree wouldn't fully avoid this either. It's possible to switch branches in a worktree. But maybe having a new (temporary) worktree makes it less likely for conflicts like this to arise? Could have this as an option.

**Comment by real-or-random** [comment]
>  I wouldn't mind some more belts and suspenders here, though I don't think it's possible to protect against a malicious parallel usage of git in general (there's always some window for a race condition).

Yep, I was more thinking of accidents -- if there's a malicious second process you're lost.

 
> A worktree wouldn't fully avoid this either. It's possible to switch branches in a worktree. But maybe having a new (temporary) worktree makes it less likely for conflicts like this to arise? Could have this as an option.

Right, this was the idea. Of course you could modify this temporary worktree in another shell but then you probably know what you're doing.

**Comment by laanwj** [comment]
If you're going to implement this, I can think of some complications:

- the user may already be using a worktree (e.g. for bitcoin core I have a (fixed) worktree per branch)
- re-using the build environment; a new worktree always starts empty, so re-bootstrapping and configure and a full rebuild is needed before testing (sometimes this may be desirable, but not always)

**Comment by MarcoFalke** [comment]
Maybe a warning that no git "write-actions" are allowed (only git "read-actions" like `show` or `diff`) would be sufficient?

**Comment by laanwj** [comment]
You could add a warning to the "documentation" I guess. But I think everyone within the very small circle of users of this tool is aware they *shouldn't* interfere with it in that way :smile: 

I think adding a check that it's signing the correct commit probably can't hurt, to prevent accidents.

**Comment by real-or-random** [comment]
> > A worktree wouldn't fully avoid this either. It's possible to switch branches in a worktree. But maybe having a new (temporary) worktree makes it less likely for conflicts like this to arise? Could have this as an option.
> 
> Right, this was the idea. Of course you could modify this temporary worktree in another shell but then you probably know what you're doing.

I've run into https://github.com/opentimestamps/opentimestamps-client/issues/87, so a worktree is not a good idea here.
